app/serializers/basic_user_serializer.rb

```diff
@@ -23,14 +23,32 @@ def user
     object[:user] || object.try(:user) || object
   end
 
+  # the scope variable passed is based on Guardian.new(current_user), and when
+  # looking at a user profile we could be looking at a different user to the
+  # current one.
+  def user_scope
```

Comment:
This worries me a bit because `BasicUserSerializer` is a superclass to so many things. It would be so easy to accidentally leak information for a user who can't see it.

I don't think we should muck with the guardian here, but instead change `notification_levels_for(scope)` to accept a guardian (scope) and a target user.

Reply:
@eviltrout I am not sure why I am following how this would leak information? The notification_levels_for methods just use scope.user and scope.anonymous. Passing in a scope and a user seems unnecessary when I could just pass in the scope for the correct user.

![image](https://user-images.githubusercontent.com/920448/118054865-6be70500-b3ca-11eb-9468-a1856f0c6a54.png)

Also all of the notification level attributes of the serializer are private, and as such are only sent back to the client if `scope.can_edit?(object)` so the risk of seeing the wrong user's stuff is low IMO:

https://github.com/discourse/discourse/blob/4923cc45fdd2076ce8043498d849db68a6d2d31b/app/serializers/user_serializer.rb#L206-L243

https://github.com/discourse/discourse/blob/4923cc45fdd2076ce8043498d849db68a6d2d31b/app/serializers/user_card_serializer.rb#L14-L23

Reply:
To be clear I agree your current implementation is safe. What concerns me is other developers coming across this API in the future and accidentally misusing it. To my knowledge, none of our core code relies on creating a `Guardian` instance for the thing you are looking at. A `Guardian` should always represent "the user performing the action."

By changing the API to something like this:

`notification_levels_for(guardian, user)`

Both parameters would be required and it would be very difficult for someone in the future to mess it up. Also, you could remove all that logic in the `BasicUserSerializer` for `user_scope`.

In the same vein, (but not nearly as important to me), it might be cleaner to do `scope.can_edit?(user)` in your `include_` methods rather than checking `is_current_user` || `admin?`

Reply:
Ah I see what you are saying now, it is more about keeping consistent. Okay I will change the `notification_levels_for` and remove the user_scope from the serializer, however I am going to leave `is_current_user` as is -- I merely refactored an existing check out into a method, and I do not want to change the meaning of it. I think it is a valid check and just replaces doing `(object.id && object.id == scope.user.try(:id))` repeatedly.

Reply:
@eviltrout I ended up just changing `notification_levels_for` to accept a user. It was entirely unnecessary for it to be accepting an instance of `Guardian`, all that it was using it for was to check `guardian.anonymous?` which is just a fancy way of saying `user.blank?`. I ripped out the other user serializer changes too. Merging today then we can make further tweaks if necessary.

Reply:
Awesome, I prefer this quite a bit. Thanks for doing the work.